lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,227 @@
+        return '
+		<script src="https://www.paypalobjects.com/webstatic/ppplus/ppplus.min.js" type="text/javascript">
+		</script>
+		<div id="ppplus">
+		</div>
+		<script type="application/javascript">		  
+		var approvalUrl = \'' . $approvalUrl . '\'; 
+		var ppp = PAYPAL.apps.PPP({
+			"approvalUrl": approvalUrl,
+			"placeholder": "ppplus",
+			"country":"' . $countryid . '" ,
+			"mode": "' . $testmode . '",
+			onContinue: function () { ppp.doCheckout(); } ,
+		});
+		</script>
+		';
+    }
+
+
+    public function process(\Aimeos\MShop\Order\Item\Iface $order, array $params = []): ?\Aimeos\MShop\Common\Helper\Form\Iface
+    {
+        return $this->getPaymentForm($order, $params);
+    }
+
+    // protected function getData(\Aimeos\MShop\Order\Item\Base\Iface $base, string $orderid, array $params): array
+    // {
+    //  return ['payerid' => "payerid-" .$this->getContext()->getUserId() ?: "payerid-" . $orderid] + parent::getData( $base , $orderid , $params);
+    // }
+
+
+    public function updateSync(
+        \Psr\Http\Message\ServerRequestInterface $request,
+        \Aimeos\MShop\Order\Item\Iface $order
+    ): \Aimeos\MShop\Order\Item\Iface {
+        try {
+            $provider = $this->getProvider();
+            $base = $this->getOrderBase($order->getBaseId());
+
+            $params = (array) $request->getAttributes() + (array) $request->getParsedBody() + (array) $request->getQueryParams();
+            $params = $this->getData($base, $order->getId(), $params);
+            $params['transactionReference'] = $this->getTransactionReference($base);
+
+            if ($this->getValue('authorize', false) && $provider->supportsCompleteAuthorize()) {
+                $response = $provider->completeAuthorize($params)->send();
+                $status = Status::PAY_AUTHORIZED;
+            } elseif ($provider->supportsCompletePurchase()) {
+                $params['PayerID'] = $request->getQueryParams()['PayerID'] ;
+
+                $response = $provider->completePurchase($params)->send();
+                $status = Status::PAY_RECEIVED;
+            } else {
+                return $order;
+            }
+
+
+            // next command that get TransactionID was $response->getTransactionId() but it doesn't work
+            if ($response->getRequest()->getTransactionId() != $order->getId()) {
+                return $order;
+            }
+
+
+            if (method_exists($response, 'isSuccessful') && $response->isSuccessful()) {
+                $order->setPaymentStatus($status);
+            } elseif (method_exists($response, 'isPending') && $response->isPending()) {
+                $order->setPaymentStatus(Status::PAY_PENDING);
+            } elseif (
+                ( method_exists($response, 'isCancelled') && $response->isCancelled() )
+                || ( !empty($response->getData()['name']) && $response->getData()['name'] == "PAYMENT_NOT_APPROVED_FOR_EXECUTION")
```

Comment:
PAYMENT_NOT_APPROVED_FOR_EXECUTION sound like it's refused by PayPal, not cancelled by the user

Reply:
If you remove this line, is the rest of the method the same as the original one and can be removed?